File: dotnet/src/Manager.Api/Features/Users/SearchUsersByName.cs
using Manager.Api.Common;
using Manager.Api.Common.Messaging;
using Manager.Api.Database;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace Manager.Api.Features.Users;

public static class SearchUsersByName
{
    public sealed record Query(string Name) : IQuery<List<Response>>;

    public sealed record Response(long Id, string Name, string Email);

    internal sealed class Handler(ApplicationDbContext db) : IQueryHandler<Query, List<Response>>
    {
        public async Task<Result<List<Response>>> Handle(Query query, CancellationToken cancellationToken)
        {
            var users = await db.Users
                .AsNoTracking()
                .Where(u => u.Name.ToLower().Contains(query.Name.ToLower()))
                .OrderBy(u => u.Id)
                .Select(u => new Response(u.Id, u.Name, u.Email))
                .ToListAsync(cancellationToken);

            return Result.Success(users);
        }
    }

    public sealed class Endpoint : IEndpoint
    {
        public void MapEndpoint(IEndpointRouteBuilder app) =>
            app.MapGet(
                    "/api/v1/users/search-by-name",
                    async (
                        [FromQuery] string name,
                        [FromServices] IQueryHandler<Query, List<Response>> handler,
                        CancellationToken cancellationToken) =>
                    {
                        var result = await handler.Handle(new Query(name), cancellationToken);
                        return result.Match(Results.Ok, CustomResults.Problem);
                    })
                .RequireAuthorization()
                .WithTags(Tags.Users);
    }
}

File: python/src/features/users/search_users_by_name.py
"""SearchUsersByName slice — search users by name contains when authenticated."""

from __future__ import annotations

from dataclasses import dataclass

from fastapi import APIRouter, Depends
from fastapi import Query as FastApiQuery
from pydantic import BaseModel, ConfigDict
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from authentication.dependencies import get_current_subject
from common.problem import problem_response
from common.result import Result
from database.models import UserModel
from database.session import get_db

router = APIRouter(prefix="/api/v1/users", tags=["Users"])


@dataclass(frozen=True, slots=True)
class Query:
    name: str


@dataclass(frozen=True, slots=True)
class Response:
    id: int
    name: str
    email: str


class SearchUsersByNameResponseItem(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: int
    name: str
    email: str


class Handler:
    def __init__(self, session: AsyncSession) -> None:
        self._session = session

    async def handle(self, query: Query) -> Result[list[Response]]:
        models = (
            (
                await self._session.execute(
                    select(UserModel)
                    .where(func.lower(UserModel.name).contains(query.name.lower()))
                    .order_by(UserModel.id)
                )
            )
            .scalars()
            .all()
        )
        users = [
            Response(id=model.id, name=model.name, email=model.email)
            for model in models
        ]
        return Result.success(users)


def _get_handler(session: AsyncSession = Depends(get_db)) -> Handler:
    return Handler(session)


@router.get("/search-by-name", response_model=None)
async def search_users_by_name(
    name: str = FastApiQuery(...),
    _: int = Depends(get_current_subject),
    handler: Handler = Depends(_get_handler),
) -> list[SearchUsersByNameResponseItem] | object:
    result = await handler.handle(Query(name=name))
    return result.match(
        on_success=lambda value: [
            SearchUsersByNameResponseItem(id=user.id, name=user.name, email=user.email)
            for user in value
        ],
        on_failure=problem_response,
    )
